File: Python/169.Majority Element.py
#找到一个数组中的主要元素(超过一半的)，假设该元素始终存在。
class Solution(object):
    def majorityElement(self, nums):
        l = len(nums)
        if l==1:    return nums[0]
        dic , l = {} , l//2
        for n in nums:
            if dic.get(n,-1)==-1:
                dic[n] = 1
            else:
                dic[n] += 1
                if dic[n]>l:
                    return n

s = Solution()
print(s.majorityElement([1]))

File: C++/Array/169.Majority Element.cpp
/*
寻找数组中一个数，该数占了数组一半以上。
*/
/*
思路：
O(N)的解决方法，从头开始计数，相同加一，不同减一，减到0换数字。
 */


#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int length = nums.size();
        if (length==0)  return 0;
        int num,count = 0;
        for(int i=0; i < nums.size(); ++i)
        {
            if(count==0)
            {
                num = nums[i];
                count = 1;
            }
            else if(nums[i]==num)   ++count;
            else    --count;
        }
        return num;
    }
};

int main()
{
	Solution s;
	vector<int> nums;
	nums.push_back(2);
	nums.push_back(1);
	nums.push_back(1);
	nums.push_back(4);
	nums.push_back(1);
	nums.push_back(1);
	cout<<s.majorityElement(nums);
}